### PR TITLE
persist staking protocol balance (before Greenland height) to DB in a…

### PR DESCRIPTION
**Pull request by lzxm160**
…rchive mode 2440
work on #2440
based on v1.1.1
need to inject staking action with minicluster and test later

### CODE REVIEW

File: action/protocol/staking/protocol.go [hunk 1]
```diff
@@ -214,6 +217,9 @@ func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager
 	bcCtx := protocol.MustGetBlockchainCtx(ctx)
 	blkCtx := protocol.MustGetBlockCtx(ctx)
 	hu := config.NewHeightUpgrade(&bcCtx.Genesis)
+	if p.enableArchiveMode && p.hu.IsPost(config.Fairbank, blkCtx.BlockHeight) && p.hu.IsPre(config.Greenland, blkCtx.BlockHeight) {
```
Comment (lzxm160) [inline action/protocol/staking/protocol.go:220]: not sure this or just `p.enableArchiveMode && p.hu.IsPost(config.Fairbank, blkCtx.BlockHeight)`?

Comment (dustinxie) [reply action/protocol/staking/protocol.go:220]: i think should be p.enableArchiveMode && p.hu.IsPre(Greenland)

we switch to nsV2 at fairbank height, but some buckets are already created before fairbank, and as long as there is bucket, we need to save staking protocol balance

File: action/protocol/staking/protocol.go [hunk 2]
```diff
@@ -249,6 +255,20 @@ func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager
...
+func (p *Protocol) saveStakingAddressHistory(height uint64, sm protocol.StateManager) {
+	csr, err := ConstructBaseView(sm)
+	if err != nil {
+		return
+	}
+	balance := csr.BaseView().bucketPool.total
+	if balance.amount.Sign() <= 0 {
+		return
+	}
+	hei := []byte(fmt.Sprintf("%d", height))
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:267]: suggest to use byteutil.Uint64ToBytesBigEndian(height)

File: action/protocol/staking/protocol.go [hunk 3]
```diff
@@ -214,6 +217,9 @@ func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager
 	bcCtx := protocol.MustGetBlockchainCtx(ctx)
 	blkCtx := protocol.MustGetBlockCtx(ctx)
 	hu := config.NewHeightUpgrade(&bcCtx.Genesis)
+	if p.enableArchiveMode && p.hu.IsPost(config.Fairbank, blkCtx.BlockHeight) && p.hu.IsPre(config.Greenland, blkCtx.BlockHeight) {
+		p.saveStakingAddressHistory(blkCtx.BlockHeight, sm)
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:221]: also, i think should pass blkCtx.BlockHeight - 1?

for example BlockHeight = 5,500,001, this is before running this block, so at this time it reflects the balance at last block height 5,500,000

File: action/protocol/staking/protocol.go [hunk 4]
```diff
@@ -72,6 +73,7 @@ type (
 		hu                 config.HeightUpgrade
 		candBucketsIndexer *CandidatesBucketsIndexer
 		voteReviser        *VoteReviser
+		enableArchiveMode  bool
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:76]: nit: maybe just archiveMode?

File: action/protocol/staking/protocol.go [hunk 5]
```diff
@@ -249,6 +255,20 @@ func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager
 	return p.handleStakingIndexer(rp.GetEpochHeight(currentEpochNum-1), sm)
 }
 
+func (p *Protocol) saveStakingAddressHistory(height uint64, sm protocol.StateManager) {
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:258]: return error?

File: action/protocol/staking/protocol.go [hunk 6]
```diff
@@ -442,7 +462,11 @@ func (p *Protocol) ReadState(ctx context.Context, sr protocol.StateReader, metho
 	case iotexapi.ReadStakingDataMethod_CANDIDATE_BY_ADDRESS:
 		resp, height, err = readStateCandidateByAddress(ctx, csr, r.GetCandidateByAddress())
 	case iotexapi.ReadStakingDataMethod_TOTAL_STAKING_AMOUNT:
-		resp, height, err = readStateTotalStakingAmount(ctx, csr, r.GetTotalStakingAmount())
+		resp, height, err = readStateTotalStakingAmountFromHeight(csr, r.GetTotalStakingAmount(), inputHeight)
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:465]: should be like below?
```
if p.archiveMode {
  resp, height, err = readStateTotalStakingAmountFromHeight(csr, r.GetTotalStakingAmount(), inputHeight)
} else {
  resp, height, err = readStateTotalStakingAmount(ctx, csr, r.GetTotalStakingAmount())
}
```

File: action/protocol/staking/protocol.go (hunk 3 above)
Comment (dustinxie) [reply action/protocol/staking/protocol.go:221]: and condition above should be
isPre(Greenland) && BlockHeight == GreenlandHeight(), so we can save the balance for GreenlandHeight - 1

File: action/protocol/staking/protocol.go (hunk 1 above)
Comment (lzxm160) [reply action/protocol/staking/protocol.go:220]: what about the balance after greenland?
I think it's p.enableArchiveMode && p.hu.IsPost(config.Fairbank-720, blkCtx.BlockHeight)

File: action/protocol/staking/protocol.go (hunk 3 above)
Comment (lzxm160) [reply action/protocol/staking/protocol.go:221]: I'm not sure,but I'll change to height-1,and then will test with mainnet

File: action/protocol/staking/protocol.go (hunk 6 above)
Comment (lzxm160) [reply action/protocol/staking/protocol.go]: I still don't kown how to get balance with height after Greenland?I think only lastest balance is saved after Greenland

File: action/protocol/staking/protocol.go [hunk 7]
```diff
@@ -442,7 +466,12 @@ func (p *Protocol) ReadState(ctx context.Context, sr protocol.StateReader, metho
 	case iotexapi.ReadStakingDataMethod_CANDIDATE_BY_ADDRESS:
 		resp, height, err = readStateCandidateByAddress(ctx, csr, r.GetCandidateByAddress())
 	case iotexapi.ReadStakingDataMethod_TOTAL_STAKING_AMOUNT:
-		resp, height, err = readStateTotalStakingAmount(ctx, csr, r.GetTotalStakingAmount())
+		if p.archiveMode && p.hu.IsPre(config.Greenland, height) {
+			resp, height, err = readStateTotalStakingAmountFromHeight(csr, r.GetTotalStakingAmount(), inputHeight)
```
Comment (yutongp) [inline action/protocol/staking/protocol.go:470]: `readStateTotalStakingAmountFromHeight` to `readStateTotalStakingAmountFromIndexer`

File: action/protocol/staking/protocol.go (hunk 6 above)
Comment (dustinxie) [reply action/protocol/staking/protocol.go]: in archive mode, SR is a `historyStateReader`, calling State() will call sf.StateAtHeight(), which is the value at the height you pass to `historyStateReader`

File: action/protocol/staking/read_state.go [hunk 8]
```diff
@@ -216,3 +229,14 @@ func getTotalStakedAmount(ctx context.Context, csr CandidateStateReader) (*big.I
 	// otherwise read from bucket pool
 	return csr.TotalStakedAmount(), nil
 }
+
+func getTotalStakedAmountFromHeight(csr CandidateStateReader, height uint64) (*big.Int, error) {
```
Comment (yutongp) [inline action/protocol/staking/read_state.go:233]: getTotalStakedAmountFromHeight -> getTotalStakedAmountFromIndexer

File: action/protocol/staking/protocol.go [hunk 9]
```diff
@@ -217,8 +218,10 @@ func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager
 	bcCtx := protocol.MustGetBlockchainCtx(ctx)
 	blkCtx := protocol.MustGetBlockCtx(ctx)
 	hu := config.NewHeightUpgrade(&bcCtx.Genesis)
-	if p.enableArchiveMode && p.hu.IsPost(config.Fairbank, blkCtx.BlockHeight) && p.hu.IsPre(config.Greenland, blkCtx.BlockHeight) {
-		p.saveStakingAddressHistory(blkCtx.BlockHeight, sm)
+	if p.archiveMode && p.hu.IsPre(config.Greenland, blkCtx.BlockHeight) {
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:221]: should be blkCtx.BlockHeight <= hu.GreenlandBlockHeight()

File: action/protocol/staking/protocol.go [hunk 10]
```diff
@@ -236,6 +239,11 @@ func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager
 	if p.candBucketsIndexer == nil {
 		return nil
 	}
+	if p.archiveMode && blkCtx.BlockHeight <= p.hu.GreenlandBlockHeight() {
+		if err := p.saveStakingAddressHistory(blkCtx.BlockHeight, sm); err != nil {
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:243]:
```go
// balance at the time of CreatePreStates() reflects last height's 
saveStakingAddressHistory(blkCtx.BlockHeight - 1, sm)
```

File: action/protocol/staking/candidate_buckets_indexer.go [hunk 11]
```diff
@@ -171,3 +173,24 @@ func (cbi *CandidatesBucketsIndexer) GetBuckets(height uint64, offset, limit uin
 	d, err := proto.Marshal(buckets)
 	return d, height, err
 }
+
+// PutStakingBalance sets staking balance
+func (cbi *CandidatesBucketsIndexer) PutStakingBalance(height uint64, total *totalAmount) error {
+	hei := byteutil.Uint64ToBytesBigEndian(height - 1)
```
Comment (dustinxie) [inline action/protocol/staking/candidate_buckets_indexer.go:179]: use `height` here, call with `height-1`, see comments below

File: state/factory/historyfactory.go [hunk 12]
```diff
@@ -46,5 +46,6 @@ func (hReader *historyStateReader) States(opts ...protocol.StateOption) (uint64,
 
 // ReadView reads the view
 func (hReader *historyStateReader) ReadView(name string) (uint64, interface{}, error) {
-	return hReader.sf.ReadView(name)
+	_, view, err := hReader.sf.ReadView(name)
+	return hReader.height, view, err
```
Comment (dustinxie) [inline state/factory/historyfactory.go:50]: rebase and revert this, the correct fix has been checked in

File: action/protocol/staking/protocol.go [hunk 13]
```diff
@@ -236,6 +239,11 @@ func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager
 	if p.candBucketsIndexer == nil {
 		return nil
 	}
+	if p.archiveMode && blkCtx.BlockHeight <= p.hu.GreenlandBlockHeight() {
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:242]: also need `&& p.candBucketsIndexer != nil`

File: state/factory/historyfactory.go (hunk 12 above)
Comment (lzxm160) [reply state/factory/historyfactory.go:50]: this pr based on v1.1.1

File: action/protocol/staking/protocol.go (hunk 13 above)
Comment (lzxm160) [reply action/protocol/staking/protocol.go:242]: see line 239

File: state/factory/historyfactory.go (hunk 12 above)
Comment (dustinxie) [reply state/factory/historyfactory.go:50]: ok, then need to cherry-pick the fix https://github.com/iotexproject/iotex-core/commit/4456bfa417108d0985a8a6e4ef23cc3bc60adc4f

File: action/protocol/staking/protocol.go [hunk 14]
```diff
@@ -249,6 +257,18 @@ func (p *Protocol) CreatePreStates(ctx context.Context, sm protocol.StateManager
...
+func (p *Protocol) saveStakingAddressHistory(height uint64, sm protocol.StateManager) error {
+	csr, err := ConstructBaseView(sm)
+	if err != nil {
+		return err
+	}
+	balance := csr.BaseView().bucketPool.total
+	if balance.amount.Sign() <= 0 {
+		return nil
+	}
+	return p.candBucketsIndexer.PutStakingBalance(height, balance)
```
Comment (dustinxie) [inline action/protocol/staking/protocol.go:269]:
```go
// balance at the time of CreatePreStates() reflects last height's balance
return p.candBucketsIndexer.PutStakingBalance(height - 1, balance)
```

**Comment by lzxm160** [comment]
test failed with branch 24406

**Pull request closed without merging**